crates/bin/prove_block/src/state_utils.rs

```diff
@@ -42,52 +37,55 @@ pub(crate) async fn get_processed_state_update(
+    let accessed_addresses: HashSet<Felt252> = get_subcalled_contracts_from_tx_traces(&traces);
 
-    let address_to_class_hash: HashMap<Felt252, Felt252> =
-        format_deployed_contracts(&state_diff, &contracts_subcalled, provider, block_id).await;
-    let address_to_nonce: HashMap<Felt252, Felt252> = format_state_update_nonces(&state_diff, transactions);
-    let class_hash_to_compiled_class_hash: HashMap<Felt252, Felt252> = format_declared_classes(&state_diff);
-    let storage_updates: HashMap<Felt252, HashMap<Felt252, Felt252>> = format_storage_updates(&state_diff);
+    let address_to_class_hash = format_deployed_contracts(&state_diff);
+    let mut class_hash_to_compiled_class_hash: HashMap<Felt252, Felt252> = format_declared_classes(&state_diff);
+    let compiled_classes = build_compiled_class_and_maybe_update_class_hash_to_compiled_class_hash(
+        provider,
+        block_id,
+        &accessed_addresses,
+        &address_to_class_hash,
+        &mut class_hash_to_compiled_class_hash,
+    )
+    .await?;
 
-    // Collect keys without consuming the HashMaps by borrowing and cloning the keys
-    let accessed_addresses: HashSet<Felt252> = address_to_class_hash
-        .keys()
-        .copied()
-        .chain(address_to_nonce.keys().cloned())
-        .chain(storage_updates.keys().cloned())
-        .collect();
-
-    FormattedStateUpdate {
-        address_to_class_hash,
-        address_to_nonce,
-        class_hash_to_compiled_class_hash,
-        storage_updates,
-        accessed_addresses,
-    }
+    Ok(FormattedStateUpdate { class_hash_to_compiled_class_hash, compiled_classes })
 }
 
-/// Format StateDiff's NonceUpdate to a HashMap<contract_address, nonce>
-fn format_state_update_nonces(
-    state_diff: &StateDiff,
-    transactions: &[InternalTransaction],
-) -> HashMap<Felt252, Felt252> {
-    let address_to_nonce: HashMap<Felt252, Felt252> = state_diff
-        .nonces
-        .iter()
-        .map(|nonce_update| {
-            // derive original nonce
-            // TODO: understand what is going on here XD
-            let num_nonce_bumps =
-                Felt252::from(transactions.iter().fold(0, |acc, tx| {
-                    acc + if tx.sender_address == Some(nonce_update.contract_address) { 1 } else { 0 }
-                }));
-            assert!(nonce_update.nonce > num_nonce_bumps);
-            let previous_nonce = nonce_update.nonce - num_nonce_bumps;
-            (nonce_update.contract_address, previous_nonce)
-        })
-        .collect();
-    address_to_nonce
+async fn build_compiled_class_and_maybe_update_class_hash_to_compiled_class_hash(
+    provider: &JsonRpcClient<HttpTransport>,
+    block_id: BlockId,
+    accessed_addresses: &HashSet<Felt252>,
+    address_to_class_hash: &HashMap<Felt252, Felt252>,
+    class_hash_to_compiled_class_hash: &mut HashMap<Felt252, Felt252>,
+) -> Result<HashMap<Felt252, GenericCasmContractClass>, Error> {
```

Comment:
Now that we moved to the latest Blockifier, don't hesitate to use types like `CompiledClassHash`, `ClassHash` etc. This will make the code cleaner and we can get the felt inside by accessing the `.0` field.